app/src/main/java/com/sun/cookbook/data/source/RecipeDataSource.kt

```diff
@@ -10,5 +10,7 @@ interface RecipeDataSource {
         fun <T> getDataRecipeDetail(idRecipe: Int, listener: OnFetchDataJsonListener<T>)
         fun <T> getDataRecipeSimilar(idRecipeSimilar: Int, listener: OnFetchDataJsonListener<T>)
         fun <T> getDataRecipeByIngredient(nameIngredient: String, listener: OnFetchDataJsonListener<T>)
+        fun <T> searchRecipeByIngredient(ingredients: String, listener: OnFetchDataJsonListener<T>)
+        fun <T> searchRecipeByName(name: String, listener: OnFetchDataJsonListener<T>)
```

Comment:
T là gì ? 
Tại sao phải để T ở đây ? 
Có gì khác nhau khi khai báo như sau 
```suggestion
        fun searchRecipeByName(name: String, listener: OnFetchDataJsonListener<MutableList<RecipeByIngredient>>)
```

Reply:
T này là kiểu generic,
Khi không biết trước kiểu dữ liệu thì sẽ dùng cái này, 
Sẽ không khác gì khi khai báo như trên nhưng sẽ có tác dụng khi ta sử dụng lại hàm này để lấy một kiểu dữ liệu khác vd như một object chẳng hạn